File: BowlerBuilder/JavaUI/src/main/java/com/neuronrobotics/bowlerbuilder/controller/TextAreaPrintStream.java
package com.neuronrobotics.bowlerbuilder.controller;

import javafx.application.Platform;
import javafx.scene.control.TextArea;

import javax.annotation.ParametersAreNonnullByDefault;
import java.io.OutputStream;

// Simple stream to append input characters to a text area
@ParametersAreNonnullByDefault
class TextAreaPrintStream extends OutputStream {

    private final TextArea textArea;

    TextAreaPrintStream(final TextArea textArea) {
        super();
        this.textArea = textArea;
    }

    @Override
    public void write(final int character) {
        Platform.runLater(() -> textArea.appendText(String.valueOf((char) character)));
    }
}

File: BowlerBuilder/JavaUI/src/main/java/com/neuronrobotics/bowlerbuilder/controller/TextAreaPrintStream.kt
/*
 * This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at http://mozilla.org/MPL/2.0/.
 */
package com.neuronrobotics.bowlerbuilder.controller

import javafx.scene.control.TextArea
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.GlobalScope
import kotlinx.coroutines.javafx.JavaFx
import kotlinx.coroutines.launch
import java.io.OutputStream

// Simple stream to append input characters to a text area
internal class TextAreaPrintStream(private val textArea: TextArea) : OutputStream() {

    override fun write(character: Int) {
        GlobalScope.launch(context = Dispatchers.JavaFx) { textArea.appendText(character.toChar().toString()) }
    }
}
